### ISSUE TITLE
Event needs to be updated after AUEN is activated to allow sending reminders for that event

### ISSUE THREAD

**Issue by joshfeck** [eventespresso/eea-automated-upcoming-event-notifications#14]
See @Pebblo's comments in this support topic:

https://eventespresso.com/topic/automated-upcoming-event-notification-add-on-not-sending/#post-301531

**Comment by Pebblo** [comment]
First, steps to reproduce.

Create 2 events on your site **without** the AUEN add-on enabled.

I named mine AUEN 1 and AUEN 2 (name them whatever you want but I'll refer to these names here) and both have a single datetime both set to to the same date - 2019-11-10 (YYYY-MM-DD)

Activate the AUEN add-on and edit **only** AUEN 1, you don't need to make any changes just update the event.

Leave AUEN 2 untouched.

Add at least 1 registration onto both of those events on the front end, those registrations need to be Approved.

Next, go to Event Espresso -> Messages -> Default message templates -> Automated Upcoming Event Notification

Enabled at least one context for the message type, both the Registrant and Event Admin will be disabled. In the sidebar of that template set the Scheduling Settings so that the message will be triggered, I have mine set to 5 days which will obviously trigger on my events right now.

Install [WP Crontrol](https://wordpress.org/plugins/wp-crontrol/).
In Tools -> Cron Events find `AHEE__EventEspresso_AutomatedEventNotifications_core_tasks_Scheduler__check` and click run now.

The expected behaviour here would be for both events to trigger the notifications, what will actually happen is AUEN 1 will trigger their messages, AUEN 2 will not.

Run the cron again, nothing will happen.

If you update AUEN 2 (again just open up the editor and save) then AUEN 2's notifications will then trigger if you run the cron again.

2 things to note if you update AUEN 2 to test this:

1. The add-on uses an extra meta value to exclude 'already notified' events from the query so to retest without new events/registrations etc, you need to remove the entries prefixed with `ee_auen_processed_` for the 2 events in extra meta.

2. To reproduce the problem with AUEN 2 you need for it to NOT have a value set for the AUEN template(s) in `esp_event_message_template` you can either remove the row manually once you know the ID of the AUEN templates or, simply reactive AUEN open/refresh the event editor for AUEN 2 and update the event, then re-activate AUEN.

----

I spent some time digging around in this so I'll add details I've found her.

The query for the above is run on the registration model and the where conditions it uses are:

```
[where] => Array
        (
            [Event.status] => Array
                (
                    [0] => IN
                    [1] => Array
                        (
                            [0] => publish
                            [1] => sold_out
                        )

                )
            [Event.Datetime.DTT_EVT_start] => Array
                (
                    [0] => BETWEEN
                    [1] => Array
                        (
                            [0] => 1573119902
                            [1] => 1573564502
                        )
                )
            [STS_ID] => RAP
            [REG_deleted] => 0
            [OR*global_conditions] => Array
                (
                    [Event.Message_Template_Group.GRP_ID] => 15
                    [Event.Message_Template_Group.GRP_ID*null] => Array
                        (
                            [0] => IS NULL
                        )
                )
        )
```

The part that appears to be causing the issue is this the`[OR*global_conditions]` conditions.

The query from the above will be something like this:

```
SELECT REG_ID AS REG_ID, ATT_ID AS ATT_ID, Registration.EVT_ID AS EVT_ID, Registration.TXN_ID AS TXN_ID  FROM  wpta_esp_registration AS Registration  LEFT JOIN wpta_posts AS Event_CPT ON Event_CPT.ID=Registration.EVT_ID LEFT JOIN wpta_esp_event_meta AS Event_Meta ON Event_CPT.ID = Event_Meta.EVT_ID  LEFT JOIN wpta_esp_datetime AS Event___Datetime ON Event___Datetime.EVT_ID=Event_CPT.ID LEFT JOIN wpta_esp_event_message_template AS Event___Event_Message_Template ON Event___Event_Message_Template.EVT_ID=Event_CPT.ID LEFT JOIN wpta_esp_message_template_group AS Event___Message_Template_Group ON Event___Message_Template_Group.GRP_ID=Event___Event_Message_Template.GRP_ID WHERE Registration.REG_deleted = 0 AND  ( (Event_CPT.post_type = 'espresso_events')) AND  ( (Event___Datetime.DTT_deleted = 0) OR Event___Datetime.DTT_ID IS NULL) AND  ( (Event___Message_Template_Group.MTP_deleted = 0) OR Event___Message_Template_Group.GRP_ID IS NULL) AND Event_CPT.post_status IN ('publish','sold_out') AND Event___Datetime.DTT_EVT_start BETWEEN '2019-11-07 09:45:02' AND '2019-11-12 13:15:02' AND Registration.STS_ID = 'RAP' AND  (Event___Message_Template_Group.GRP_ID = 15 OR Event___Message_Template_Group.GRP_ID IS NULL) GROUP BY Registration.REG_ID
```

Without updating the event with AUEN activated, `Message_Template_Group` has NO records for the event at all that's what seems to be breaking the above query. 

Is there another way we can handle this in the models?

**Comment by mnelson4** [comment]
> Is there another way we can handle this in the models to will this need a custom/multiple queries?

The above query would work if there were absolutely NO entries in the join-table between the event and ANY message template groups, or if there's an entry for message template group 15. BUT there's always a few entries (for other message template groups). There are 3 possible fixes I see:

1. In order to do this all in one database query, we'd need to modify the models to allow setting a condition on the JOIN (so we'd only join to message template groups with 15 or not at all), rather than in a WHERE condition after the fact. Or a subquery would probably do the trick. Either way, it would currently require a hand-written database query, or adding some more capabilities to the models.

2. The workaround is to instead first just query to see if there's a relation to message template group 15 from the event. eg `EEM_Message_Template_Group::instance()->exists([['GRP_ID' => 15, 'Event__EVT' => 12345]]);`

If there is no such entry, create one. I'm not sure if that's feasible. Where in the code is this query happening @Pebblo?

3. Lastly, we could have a DMS to cycle through ALL events, and make sure they have a message template group set for AUEN.

**Comment by Pebblo** [comment]
It happens in 2 places, both very similar:

Upcoming **Event** Notifications - 
https://github.com/eventespresso/eea-automated-upcoming-event-notifications/blob/master/domain/services/commands/message/UpcomingEventNotificationsCommandHandler.php#L148-L189

Upcoming **Datetime** Notifications - 
https://github.com/eventespresso/eea-automated-upcoming-event-notifications/blob/master/domain/services/commands/message/UpcomingDatetimeNotificationsCommandHandler.php#L258-L287

**Comment by Pebblo** [comment]
I get the feeling that because the add-on is pulling the registrations directly, all of the above options are going to problematic.

#1 needs to pull the registrations from events that either has the global AUEN template ID set on them, or no AUEN template set at all (meaning they should just use the global as they obviously don't have a custom template set).

#2 we'll need to query the events first to do that, in which case would it be better to query for the events and pass the ID's of those events into the registration model to pull them directly?

#3 could work, seems like a lot of work for this though? It would also need to run every time the add-on is activated as if you save an event without the add-on active, its `Message_Template_Group` rows for the AUEN templates are removed.

Things have changed with the add-on scheduling from when this was set up, originally it ran once a day so it had to do everything in one go, now it runs every 3 hours so maybe we could come at this from another angle?

**Comment by mnelson4** [comment]
# Quick Summary
I think the issue boils down to two main options:
1. Add a feature onto the models (which could get hairy, especially if we don't set a budget for the project and we end up wanting to entirely refactor the models)
2. Somehow loop over all events and make sure they have the AUEN template setup (either in a DMS, or that could also be part of a CRON task).

# More Details
I think the heart of the problem is the models can't be used to query for items that lack related item meeting certain criteria if there are other related items that don't meet the criteria- eg, it can't query for events that don't use message template group 15 when those events all use other message template groups. If we added the ability for models to add conditions onto JOINs' "on" statements, that issue would be solved, which would actually help in all three options:
1) we could successfully query for registrations for events using message template group15 or none at all
2) we could query for events with no message template group, and pass that list of events into the registrations query (although we probably wouldn't, as option 1 would do it in one query and appears pretty obviously better)
3) we could query for all events that don't use message template group 15, and add it to them, in order to avoid future monkeying-around.

Doing that will require going into the model guts and carving out the feature, and some lead developers 🙊 are chomping at the bit to refactor them (although they may also be distracted with all the JS excitement going on currently), which could make the project explode if proper limitations on the project aren't established.

FYI I confirmed adding `AND Event___Event_Message_Template.GRP_ID = 15` right after `LEFT JOIN wp_esp_event_message_template AS Event___Event_Message_Template ON Event___Event_Message_Template.EVT_ID=Event_CPT.ID` seems to make the query work as intended. So that's what we'd want to achieve for option 1. 

I think that would be a useful feature for the models. It could look something like

```
$registrations = EEM_Registration::instance()->get_all(
    'on_join_conditions' => [
        'Event__Event_Message_Template__GRP_ID' => 15
    ]
);
```
So when it's doing the join to the message template group join-table, it would add `AND Event___Event_Message_Template.GRP_ID = 15` onto the join condition.

I think if we don't add that feature to the models, we're kinda forced to need to loop over all events and make sure they have the AUEN message template group setup (during a DMS, which was option 3; or maybe we could do it on a cron task).

Thoughts @Pebblo or @joshfeck? I think deciding which direction to attempt is the main next step.

**Comment by joshfeck** [comment]
I'm thinking this is a wontfix if the above are the only options. We can add a note in the documentation that says any existing events will need to be updated before they'll be added to the Upcoming events notifications queue. That's a one time task for any new users of the add-on.

I'd be open to discuss adding a script to an activation hook where on plugin activate the Message_Template_Group relationship could be added to all upcoming events. Or if need be a utility (like we have for resetting capabilities) where the site admin could go to the EE > Maintenance page, hit a button, and the Message_Template_Group relationship could be added to all upcoming events.

**Comment by joshfeck** [comment]
@mnelson4 Like Tony already alluded to, could we write a custom query (with good old $wpdb) and just not use the models system 🙊 for that task?

**Comment by mnelson4** [comment]
> I'm thinking this is a wontfix
👏 for saying that! But we'll see if it comes to that...

> script to an activation hook where on plugin activate the Message_Template_Group relationship could be added to all upcoming events

That's kinda what I was suggesting with a DMS. I suggested a DMS though in case they have thousands of upcoming events, in which case the request could time out etc.

> Or if need be a utility (like we have for resetting capabilities) where the site admin could go to the EE > Maintenance page, hit a button, and the Message_Template_Group relationship could be added to all upcoming events.

Yeah that would work too. It could be a batch process in order to avoid timeouts again. pOtatoe, poTAto.

>  could we write a custom query (with good old $wpdb) and just not use the models system speak_no_evilfor that task?

Yeah I looked into that, (and it didn't end up in my last reply) and I actually got halfway through doing that as an experiment. The main difficulty is that the functions in question accept an array of arbitrary `$query_params` from at least a few function calls upstream. I don't think writing custom code to parse through that to create the needed SQL isn't feasible, so we would need to instead refactor the upstream code to only pass something simpler.
Other than that, yes we can just copy-and-paste the [SQL tony already pasted above](https://github.com/eventespresso/eea-automated-upcoming-event-notifications/issues/14#issuecomment-550243004), and tweak it to be what we need.

So it would require a little bit of refactoring the upstream AUEN code (not just [the functions performing the queries that tony pointed to](https://github.com/eventespresso/eea-automated-upcoming-event-notifications/issues/14#issuecomment-550481004)), but would work. (I think it's fairly appropriate for this situation, but iirc this isn't the first time we've had an issue like this and resorted to a manual query because we couldn't add conditions onto the joins.)

**Comment by mnelson4** [comment]
So @joshfeck would you like @Pebblo or I to attempt to rewrite the query in raw MySQL and do the refactoring so not pass in `$additional_where_params` to the refactor methods? I would guess it would be 2-4 hours of initial development before it could be tested

**Comment by joshfeck** [comment]
Yes that would be fine to 
1) use a custom query
2) simplify the upstream AUEN code

users with sites that have a lot of existing upcoming events would appreciate not having to resave every event in order to get notifications working.

**Comment by joshfeck** [comment]
fixed in #15
